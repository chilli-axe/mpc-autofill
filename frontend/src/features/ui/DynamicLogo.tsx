import Image from "next/image";
import React, { useEffect, useState } from "react";
import Col from "react-bootstrap/Col";
import Row from "react-bootstrap/Row";
import styled, { css, keyframes, StyledComponent } from "styled-components";

import { QueryTags } from "@/common/constants";
import { CardType } from "@/common/schema_types";
import { CardDocument, useAppDispatch } from "@/common/types";
import { Spinner } from "@/components/Spinner";
import {
  MemoizedCardImage,
  MemoizedCardProportionWrapper,
} from "@/features/card/Card";
import { lato } from "@/pages/_app";
import { api, useGetSampleCardsQuery } from "@/store/api";
import {
  useBackendConfigured,
  useProjectName,
} from "@/store/slices/backendSlice";

//# region styled components

// note: i'm using transform3d rather than transform bc it apparently fixes animations on ios safari
// https://www.reddit.com/r/webdev/comments/iv26y7/iossafari_css_animation_not_working/

const DynamicLogoContainer = styled.div`
  position: relative;
  width: 85%;
  left: 7.5%;
  aspect-ratio: 1 / 1;
  background: linear-gradient(#4692f0, #183251);
  border-radius: 50%;
  // would like to do the first one of these with outline, but alas
  // all my homies hate safari
  box-shadow: 0 0 0 4px black, 0 1rem 2rem rgba(0, 0, 0, 0.4);
`;

const DynamicLogoLabel = styled.p`
  position: absolute;
  font-weight: bold;
  top: 19.3548%;
  left: 50%;
  z-index: 10;
  transform: translate3d(-50%, 0, 0);
  // TODO: find a better implementation of font scaling
  font-size: min(4em, 10vw);
  text-shadow: 0 4px 15px #000000;
  white-space: nowrap;
  user-select: none;
  color: white;
`;

const DynamicLogoArrowKeyframes = keyframes`
  0% {
    transform: translate3d(-50%, 16%, 0);
    opacity: 0;
  }
  100% {
    transform: translate3d(-50%, 0, 0);
    opacity: 1;
  }
`;

const DynamicLogoArrowWrapper = styled.div<{ animated?: boolean }>`
  width: 40%;
  height: 40%;
  position: absolute;
  left: 50%;
  transform: translate3d(-50%, -16%, 0);
  opacity: 0;
  ${(props) =>
    props.animated === true &&
    css`
      animation: ${DynamicLogoArrowKeyframes} 0.75s ease-out 0.01s forwards;
    `}
`;

const ImageTransformWrapperBase = styled.div<{
  animated?: boolean;
}>`
  position: absolute;
  height: 32.358%;
  width: 32.358%;
  left: 33.871%;
  top: 33.871%;
  transition-property: height, width, top, left, z-index;
  transition-duration: 0.15s;
  transition-timing-function: ease-in-out;
  &:hover {
    height: 40%;
    width: 40%;
    top: 20%;
    left: 30%;
    z-index: 20;
  }
`;

const FirstImageTransformKeyframes = keyframes`
  0% {
    transform: translate3d(0, 0, 0);
    rotate: 0;
  }
  100% {
    transform: translate3d(-120%, -27%, 0);
    rotate: -40deg;
  }
`;

const FirstImageTransformWrapper = styled(ImageTransformWrapperBase)<{
  animated?: boolean;
}>`
  z-index: 0;
  ${(props) =>
    props.animated === true &&
    css`
      animation: ${FirstImageTransformKeyframes} 1s ease-in-out 0.01s forwards;
    `}
`;

const SecondImageTransformKeyframes = keyframes`
  0% {
    transform: translate3d(0, 0, 0);
    rotate: 0;
  }
  100% {
    transform: translate3d(-60%, 0, 0);
    rotate: -20deg;
  }
`;

const SecondImageTransformWrapper = styled(ImageTransformWrapperBase)<{
  animated?: boolean;
}>`
  z-index: 1;
  ${(props) =>
    props.animated === true &&
    css`
      animation: ${SecondImageTransformKeyframes} 1s ease-in-out 0.01s forwards;
    `}
`;

const ThirdImageTransformKeyframes = keyframes`
  0% {
    transform: translate3d(0, 0, 0);
  }
  100% {
    transform: translate3d(0, 6%, 0);
  }
`;

const ThirdImageTransformWrapper = styled(ImageTransformWrapperBase)<{
  animated?: boolean;
}>`
  z-index: 2;
  ${(props) =>
    props.animated === true &&
    css`
      animation: ${ThirdImageTransformKeyframes} 1s ease-in-out 0.01s forwards;
    `}
`;

const FourthImageTransformKeyframes = keyframes`
  0% {
    transform: translate3d(0, 0, 0);
    rotate: 0;
  }
  100% {
    transform: translate3d(60%, 0, 0);
    rotate: 20deg;
  }
`;

const FourthImageTransformWrapper = styled(ImageTransformWrapperBase)<{
  animated?: boolean;
}>`
  z-index: 3;
  ${(props) =>
    props.animated === true &&
    css`
      animation: ${FourthImageTransformKeyframes} 1s ease-in-out 0.01s forwards;
    `}
`;

const FifthImageTransformKeyframes = keyframes`
  0% {
    transform: translate3d(0, 0, 0);
    rotate: 0;
  }
  100% {
    transform: translate3d(120%, -27%, 0);
    rotate: 40deg;
  }
`;

const FifthImageTransformWrapper = styled(ImageTransformWrapperBase)<{
  animated?: boolean;
}>`
  z-index: 4;
  ${(props) =>
    props.animated === true &&
    css`
      animation: ${FifthImageTransformKeyframes} 1s ease-in-out 0.01s forwards;
    `}
`;

//# endregion

const SampleCardDocument: CardDocument = {
  identifier: "your-design-here",
  cardType: CardType.Card,
  name: "Your Design Here",
  priority: 0,
  source: "",
  sourceName: "",
  sourceId: 0,
  sourceVerbose: "",
  sourceType: undefined,
  sourceExternalLink: undefined,
  dpi: 300,
  searchq: "",
  extension: "png",
  dateCreated: "1st January, 2000",
  dateModified: "1st January, 2000",
  downloadLink: "",
  size: 1,
  smallThumbnailUrl: "/logo-blank.png",
  mediumThumbnailUrl: "",
  language: "EN",
  tags: [],
};

export function DynamicLogo() {
  //# region hooks and queries

  const dispatch = useAppDispatch();
  const backendConfigured = useBackendConfigured();
  const projectName = useProjectName();
  const sampleCardsQuery = useGetSampleCardsQuery();

  //# endregion

  //# region state

  const [loading, setLoading] = useState<boolean>(true);

  //# endregion

  //# region effects

  useEffect(() => {
    dispatch(api.util.invalidateTags([QueryTags.SampleCards]));
  }, [dispatch]);
  // this ignores the initial flash of styled-components not doing the thing on first page load
  useEffect(() => setLoading(false), []);

  //# endregion

  //# region computed constants

  const displayCards: Array<
    [
      CardDocument | null,
      StyledComponent<typeof ImageTransformWrapperBase, any>
    ]
  > = [
    [
      sampleCardsQuery.isSuccess
        ? sampleCardsQuery.data["TOKEN"][0] ??
          sampleCardsQuery.data["CARDBACK"][0]
        : null,
      FirstImageTransformWrapper,
    ],
    [
      sampleCardsQuery.isSuccess ? sampleCardsQuery.data["CARD"][0] : null,
      SecondImageTransformWrapper,
    ],
    [
      sampleCardsQuery.isSuccess ? sampleCardsQuery.data["CARD"][1] : null,
      ThirdImageTransformWrapper,
    ],
    [
      sampleCardsQuery.isSuccess ? sampleCardsQuery.data["CARD"][2] : null,
      FourthImageTransformWrapper,
    ],
    [
      sampleCardsQuery.isSuccess ? sampleCardsQuery.data["CARD"][3] : null,
      FifthImageTransformWrapper,
    ],
  ];
  const animated = !backendConfigured || sampleCardsQuery.isSuccess;

  //# endregion

  return (
    <>
      {loading ? (
        <Spinner size={12} />
      ) : (
        <Row className="justify-content-center" data-testid="dynamic-logo">
          <Col xl={6} lg={7} md={8} sm={12} xs={12}>
            <DynamicLogoContainer>
              <DynamicLogoLabel className={lato.className}>
                {projectName}
              </DynamicLogoLabel>
              <DynamicLogoArrowWrapper animated={animated}>
                <Image
                  src="/arrow.svg"
                  alt="logo-arrow"
                  quality={100}
                  fill={true}
                />
              </DynamicLogoArrowWrapper>

              {displayCards.map(
                ([maybeCardDocument, WrapperElement], index) => (
                  <WrapperElement
                    key={`logo-card${index}-outer-wrapper`}
                    animated={animated}
                  >
                    <MemoizedCardProportionWrapper
                      bordered={!backendConfigured}
                      small={true}
                      key={`$logo-card${index}-inner-wrapper`}
                    >
                      <MemoizedCardImage
                        key={`$logo-card${index}-image`}
                        maybeCardDocument={
                          backendConfigured
                            ? maybeCardDocument
                            : SampleCardDocument
                        }
                        hidden={false}
                        small={true}
                        showDetailedViewOnClick={backendConfigured}
                      />
                    </MemoizedCardProportionWrapper>
                  </WrapperElement>
                )
              )}
            </DynamicLogoContainer>
          </Col>
        </Row>
      )}
    </>
  );
}
