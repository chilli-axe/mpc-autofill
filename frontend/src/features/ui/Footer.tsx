import Link from "next/link";

import { useGetBackendInfoQuery } from "@/store/api";

function Spacer() {
  return (
    <span style={{ marginLeft: 0.25 + "em", marginRight: 0.25 + "em" }}> </span>
  );
}

export default function Footer() {
  const backendInfoQuery = useGetBackendInfoQuery();
  return (
    <>
      <hr />
      <footer className="page-footer font-small blue">
        <div className="footer-copyright text-center py-3">
          Made with ♥️ by chilli_axe
          <Spacer />•<Spacer />
          <a href="https://github.com/chilli-axe/mpc-autofill" target="_blank">
            GitHub
          </a>
          {backendInfoQuery.isSuccess && backendInfoQuery.data?.reddit != null && (
            <>
              <Spacer />•<Spacer />
              <a href={backendInfoQuery.data.reddit} target="_blank">
                Reddit
              </a>
            </>
          )}
          {backendInfoQuery.isSuccess &&
            backendInfoQuery.data?.discord != null && (
              <>
                <Spacer />•<Spacer />
                <a href={backendInfoQuery.data.discord} target="_blank">
                  Discord
                </a>
              </>
            )}
          <Spacer />•<Spacer />
          <Link href="/about">About</Link>
        </div>
      </footer>
    </>
  );
}
