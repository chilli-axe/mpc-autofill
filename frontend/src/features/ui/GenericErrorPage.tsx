import styled from "styled-components";

const TextWrapper = styled.div`
  position: absolute;
  top: 50%;
  left: 50%;
  transform: translate(-50%, -50%);
  text-align: center;
`;
const HeaderText = styled.h1`
  position: relative;
`;
export const LargerText = styled.p`
  font-size: 1.25em;
  position: relative;
`;

interface ErrorPageProps {
  title: string;
  text: Array<string>;
}

export function GenericErrorPage({ title, text }: ErrorPageProps) {
  return (
    <TextWrapper>
      <h1>{title}</h1>
      {text.map((line, index) => (
        <LargerText key={index.toString()}>{line}</LargerText>
      ))}
    </TextWrapper>
  );
}
