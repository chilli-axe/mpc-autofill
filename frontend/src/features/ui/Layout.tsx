import { Queue } from "async-await-queue";
import { useRouter } from "next/router";
import { GoogleAnalytics } from "nextjs-google-analytics";
import React, { useEffect } from "react";
import { PropsWithChildren } from "react";
import Container from "react-bootstrap/Container";
import SSRProvider from "react-bootstrap/SSRProvider";
import { Provider } from "react-redux";
import styled from "styled-components";

import { ContentMaxWidth, NavbarHeight } from "@/common/constants";
import {
  getGoogleAnalyticsConsent,
  getLocalStorageBackendURL,
  setLocalStorageBackendURL,
} from "@/common/cookies";
import { standardiseURL } from "@/common/processing";
import { useAppDispatch, useAppSelector } from "@/common/types";
import {
  DownloadContext,
  DownloadContextProvider,
} from "@/features/download/download";
import { Modals } from "@/features/modals/Modals";
import { Toasts } from "@/features/toasts/Toasts";
import ProjectNavbar from "@/features/ui/Navbar";
import {
  selectBackendURL,
  setURL,
  useBackendConfigured,
} from "@/store/slices/backendSlice";
import store from "@/store/store";

function BackendSetter() {
  const router = useRouter();
  const { server } = router.query;
  const formattedURL: string | null =
    server != null && typeof server == "string" && server.length > 0
      ? standardiseURL(server.trim())
      : null;

  const dispatch = useAppDispatch();
  const backendConfigured = useBackendConfigured();
  const backendURL = useAppSelector(selectBackendURL);
  useEffect(() => {
    const envURL = process.env.NEXT_PUBLIC_BACKEND_URL;
    const localStorageBackendURL = getLocalStorageBackendURL();
    if (
      localStorageBackendURL != undefined &&
      backendURL !== localStorageBackendURL
    ) {
      // TODO: stale value here
      dispatch(setURL(localStorageBackendURL));
    } else if (envURL != null && envURL !== localStorageBackendURL) {
      setLocalStorageBackendURL(envURL);
      if (!backendConfigured) {
        dispatch(setURL(envURL));
      }
    } else if (formattedURL != null) {
      dispatch(setURL(formattedURL));
      setLocalStorageBackendURL(formattedURL);
      if (server != null && typeof server == "string" && server.length > 0) {
        // @ts-ignore  // TODO
        router.replace({ server }, undefined, { shallow: true });
      }
    }
  }, [router.isReady, backendConfigured, formattedURL, dispatch, backendURL]);

  return <></>;
}

const OverscrollProvider = styled(Provider)`
  overscroll-behavior: none;
  overflow-x: hidden;
  overflow-y: hidden; // https://stackoverflow.com/a/69589919/13021511
`;

const ContentContainer = styled(Container)`
  overflow-y: scroll;
  overflow-x: hidden;
  top: ${NavbarHeight}px;
  position: fixed;
  height: calc(
    100vh - ${NavbarHeight}px
  ); // for compatibility with older browsers
  height: calc(100dvh - ${NavbarHeight}px); // handles the ios address bar
`;

const MaxWidthContainer = styled(Container)`
  max-width: ${ContentMaxWidth}px;
`;

interface ProjectContainerProps {
  gutter?: number;
}

export function ProjectContainer({
  gutter = 2,
  children,
}: PropsWithChildren<ProjectContainerProps>) {
  return (
    <ContentContainer fluid className={`g-${gutter}`}>
      <MaxWidthContainer className={`g-${gutter}`}>
        {children}
      </MaxWidthContainer>
    </ContentContainer>
  );
}

export function LayoutWithoutReduxProvider({ children }: PropsWithChildren) {
  const consent = getGoogleAnalyticsConsent();
  const downloadContext: DownloadContext = new Queue(10, 50);
  return (
    <DownloadContextProvider value={downloadContext}>
      {consent === true && (
        <GoogleAnalytics trackPageViews gaMeasurementId="G-JV8WV3FQML" />
      )}
      <Toasts />
      <Modals />
      <BackendSetter />
      <ProjectNavbar />
      {children}
    </DownloadContextProvider>
  );
}

export default function Layout({ children }: PropsWithChildren) {
  return (
    <>
      <SSRProvider>
        <OverscrollProvider store={store}>
          <LayoutWithoutReduxProvider>{children}</LayoutWithoutReduxProvider>
        </OverscrollProvider>
      </SSRProvider>
    </>
  );
}
