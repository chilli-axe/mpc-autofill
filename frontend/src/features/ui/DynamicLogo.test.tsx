import { screen, waitFor } from "@testing-library/react";

import { cardDocument1, noBackend } from "@/common/test-constants";
import { renderWithProviders } from "@/common/test-utils";
import { DynamicLogo } from "@/features/ui/DynamicLogo";
import { defaultHandlers } from "@/mocks/handlers";
import { server } from "@/mocks/server";

test("the html structure of the dynamic logo, backend configured", async () => {
  server.use(...defaultHandlers);
  const rendered = renderWithProviders(<DynamicLogo />);

  await waitFor(() =>
    expect(screen.getByAltText(cardDocument1.name)).toBeInTheDocument()
  );
  expect(rendered.getByTestId("dynamic-logo")).toMatchSnapshot();
});

test("the html structure of the dynamic logo, no backend configured", async () => {
  server.use(...defaultHandlers);
  const rendered = renderWithProviders(
    <DynamicLogo />,
    {
      preloadedState: { backend: noBackend },
    },
    false
  );

  await waitFor(() =>
    expect(screen.getAllByAltText("Your Design Here")).toHaveLength(5)
  );
  expect(rendered.getByTestId("dynamic-logo")).toMatchSnapshot();
});
