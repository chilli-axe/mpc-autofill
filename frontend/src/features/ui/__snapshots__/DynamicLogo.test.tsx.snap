// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the html structure of the dynamic logo, backend configured 1`] = `
.c6 {
  z-index: 1;
  opacity: 1;
}

.c6:hover {
  cursor: pointer;
}

.c5 {
  z-index: 0;
  background: #4e5d6c;
  border: solid 0px black;
}

.c0 {
  position: relative;
  width: 85%;
  left: 7.5%;
  aspect-ratio: 1 / 1;
  background: linear-gradient(#4692f0,#183251);
  border-radius: 50%;
  box-shadow: 0 0 0 4px black,0 1rem 2rem rgba(0,0,0,0.4);
}

.c1 {
  position: absolute;
  font-weight: bold;
  top: 19.3548%;
  left: 50%;
  z-index: 10;
  -webkit-transform: translate3d(-50%,0,0);
  -ms-transform: translate3d(-50%,0,0);
  transform: translate3d(-50%,0,0);
  font-size: min(4em,10vw);
  text-shadow: 0 4px 15px #000000;
  white-space: nowrap;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  color: white;
}

.c2 {
  width: 40%;
  height: 40%;
  position: absolute;
  left: 50%;
  -webkit-transform: translate3d(-50%,-16%,0);
  -ms-transform: translate3d(-50%,-16%,0);
  transform: translate3d(-50%,-16%,0);
  opacity: 0;
  -webkit-animation: eOZAMF 0.75s ease-out 0.01s forwards;
  animation: eOZAMF 0.75s ease-out 0.01s forwards;
}

.c3 {
  position: absolute;
  height: 32.358%;
  width: 32.358%;
  left: 33.871%;
  top: 33.871%;
  -webkit-transition-property: height,width,top,left,z-index;
  transition-property: height,width,top,left,z-index;
  -webkit-transition-duration: 0.15s;
  transition-duration: 0.15s;
  -webkit-transition-timing-function: ease-in-out;
  transition-timing-function: ease-in-out;
}

.c3:hover {
  height: 40%;
  width: 40%;
  top: 20%;
  left: 30%;
  z-index: 20;
}

.c4 {
  z-index: 0;
  -webkit-animation: bGJjOD 1s ease-in-out 0.01s forwards;
  animation: bGJjOD 1s ease-in-out 0.01s forwards;
}

.c7 {
  z-index: 1;
  -webkit-animation: oyYHp 1s ease-in-out 0.01s forwards;
  animation: oyYHp 1s ease-in-out 0.01s forwards;
}

.c8 {
  z-index: 2;
  -webkit-animation: hNfKvJ 1s ease-in-out 0.01s forwards;
  animation: hNfKvJ 1s ease-in-out 0.01s forwards;
}

.c9 {
  z-index: 3;
  -webkit-animation: fZMtLp 1s ease-in-out 0.01s forwards;
  animation: fZMtLp 1s ease-in-out 0.01s forwards;
}

.c10 {
  z-index: 4;
  -webkit-animation: wzgiT 1s ease-in-out 0.01s forwards;
  animation: wzgiT 1s ease-in-out 0.01s forwards;
}

<div
  class="justify-content-center row"
  data-testid="dynamic-logo"
>
  <div
    class="col-xl-6 col-lg-7 col-md-8 col-sm-12 col-12"
  >
    <div
      class="c0"
    >
      <p
        class="c1 className"
      >
        Test Site
      </p>
      <div
        class="c2"
      >
        <img
          alt="logo-arrow"
          data-nimg="fill"
          decoding="async"
          loading="lazy"
          src="/arrow.svg"
          style="position: absolute; height: 100%; width: 100%; left: 0px; top: 0px; right: 0px; bottom: 0px; color: transparent;"
        />
      </div>
      <div
        class="c3 c4"
      >
        <div
          class="c5 rounded-lg shadow-lg ratio ratio-7x5"
        >
          <img
            alt="Card 6"
            class="c6 card-img card-img-fade-in"
            data-nimg="fill"
            decoding="async"
            loading="lazy"
            src=""
            style="position: absolute; height: 100%; width: 100%; left: 0px; top: 0px; right: 0px; bottom: 0px; color: transparent;"
          />
        </div>
      </div>
      <div
        class="c3 c7"
      >
        <div
          class="c5 rounded-lg shadow-lg ratio ratio-7x5"
        >
          <img
            alt="Card 1"
            class="c6 card-img card-img-fade-in"
            data-nimg="fill"
            decoding="async"
            loading="lazy"
            src=""
            style="position: absolute; height: 100%; width: 100%; left: 0px; top: 0px; right: 0px; bottom: 0px; color: transparent;"
          />
        </div>
      </div>
      <div
        class="c3 c8"
      >
        <div
          class="c5 rounded-lg shadow-lg ratio ratio-7x5"
        >
          <img
            alt="Card 2"
            class="c6 card-img card-img-fade-in"
            data-nimg="fill"
            decoding="async"
            loading="lazy"
            src=""
            style="position: absolute; height: 100%; width: 100%; left: 0px; top: 0px; right: 0px; bottom: 0px; color: transparent;"
          />
        </div>
      </div>
      <div
        class="c3 c9"
      >
        <div
          class="c5 rounded-lg shadow-lg ratio ratio-7x5"
        >
          <img
            alt="Card 3"
            class="c6 card-img card-img-fade-in"
            data-nimg="fill"
            decoding="async"
            loading="lazy"
            src=""
            style="position: absolute; height: 100%; width: 100%; left: 0px; top: 0px; right: 0px; bottom: 0px; color: transparent;"
          />
        </div>
      </div>
      <div
        class="c3 c10"
      >
        <div
          class="c5 rounded-lg shadow-lg ratio ratio-7x5"
        >
          <img
            alt="Card 4"
            class="c6 card-img card-img-fade-in"
            data-nimg="fill"
            decoding="async"
            loading="lazy"
            src=""
            style="position: absolute; height: 100%; width: 100%; left: 0px; top: 0px; right: 0px; bottom: 0px; color: transparent;"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`the html structure of the dynamic logo, no backend configured 1`] = `
.c6 {
  width: 4em;
  height: 4em;
  z-index: 2;
}

.c7 {
  z-index: 1;
  opacity: 0;
}

.c7:hover {
  cursor: auto;
}

.c5 {
  z-index: 0;
  background: #4e5d6c;
  border: solid 2px black;
}

.c0 {
  position: relative;
  width: 85%;
  left: 7.5%;
  aspect-ratio: 1 / 1;
  background: linear-gradient(#4692f0,#183251);
  border-radius: 50%;
  box-shadow: 0 0 0 4px black,0 1rem 2rem rgba(0,0,0,0.4);
}

.c1 {
  position: absolute;
  font-weight: bold;
  top: 19.3548%;
  left: 50%;
  z-index: 10;
  -webkit-transform: translate3d(-50%,0,0);
  -ms-transform: translate3d(-50%,0,0);
  transform: translate3d(-50%,0,0);
  font-size: min(4em,10vw);
  text-shadow: 0 4px 15px #000000;
  white-space: nowrap;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  color: white;
}

.c2 {
  width: 40%;
  height: 40%;
  position: absolute;
  left: 50%;
  -webkit-transform: translate3d(-50%,-16%,0);
  -ms-transform: translate3d(-50%,-16%,0);
  transform: translate3d(-50%,-16%,0);
  opacity: 0;
  -webkit-animation: eOZAMF 0.75s ease-out 0.01s forwards;
  animation: eOZAMF 0.75s ease-out 0.01s forwards;
}

.c3 {
  position: absolute;
  height: 32.358%;
  width: 32.358%;
  left: 33.871%;
  top: 33.871%;
  -webkit-transition-property: height,width,top,left,z-index;
  transition-property: height,width,top,left,z-index;
  -webkit-transition-duration: 0.15s;
  transition-duration: 0.15s;
  -webkit-transition-timing-function: ease-in-out;
  transition-timing-function: ease-in-out;
}

.c3:hover {
  height: 40%;
  width: 40%;
  top: 20%;
  left: 30%;
  z-index: 20;
}

.c4 {
  z-index: 0;
  -webkit-animation: bGJjOD 1s ease-in-out 0.01s forwards;
  animation: bGJjOD 1s ease-in-out 0.01s forwards;
}

.c8 {
  z-index: 1;
  -webkit-animation: oyYHp 1s ease-in-out 0.01s forwards;
  animation: oyYHp 1s ease-in-out 0.01s forwards;
}

.c9 {
  z-index: 2;
  -webkit-animation: hNfKvJ 1s ease-in-out 0.01s forwards;
  animation: hNfKvJ 1s ease-in-out 0.01s forwards;
}

.c10 {
  z-index: 3;
  -webkit-animation: fZMtLp 1s ease-in-out 0.01s forwards;
  animation: fZMtLp 1s ease-in-out 0.01s forwards;
}

.c11 {
  z-index: 4;
  -webkit-animation: wzgiT 1s ease-in-out 0.01s forwards;
  animation: wzgiT 1s ease-in-out 0.01s forwards;
}

<div
  class="justify-content-center row"
  data-testid="dynamic-logo"
>
  <div
    class="col-xl-6 col-lg-7 col-md-8 col-sm-12 col-12"
  >
    <div
      class="c0"
    >
      <p
        class="c1 className"
      >
        MPC Autofill
      </p>
      <div
        class="c2"
      >
        <img
          alt="logo-arrow"
          data-nimg="fill"
          decoding="async"
          loading="lazy"
          src="/arrow.svg"
          style="position: absolute; height: 100%; width: 100%; left: 0px; top: 0px; right: 0px; bottom: 0px; color: transparent;"
        />
      </div>
      <div
        class="c3 c4"
      >
        <div
          class="c5 rounded-lg shadow-lg ratio ratio-7x5"
        >
          <div
            class="d-flex justify-content-center align-items-center"
          >
            <div
              class="c6 spinner-border"
              role="status"
              size="4"
            >
              <span
                class="visually-hidden"
              >
                Loading...
              </span>
            </div>
          </div>
          <img
            alt="Your Design Here"
            class="c7 card-img card-img-fade-in"
            data-nimg="fill"
            decoding="async"
            loading="lazy"
            sizes="100vw"
            src="/_next/image?url=%2Flogo-blank.png&w=3840&q=75"
            srcset="/_next/image?url=%2Flogo-blank.png&w=640&q=75 640w, /_next/image?url=%2Flogo-blank.png&w=750&q=75 750w, /_next/image?url=%2Flogo-blank.png&w=828&q=75 828w, /_next/image?url=%2Flogo-blank.png&w=1080&q=75 1080w, /_next/image?url=%2Flogo-blank.png&w=1200&q=75 1200w, /_next/image?url=%2Flogo-blank.png&w=1920&q=75 1920w, /_next/image?url=%2Flogo-blank.png&w=2048&q=75 2048w, /_next/image?url=%2Flogo-blank.png&w=3840&q=75 3840w"
            style="position: absolute; height: 100%; width: 100%; left: 0px; top: 0px; right: 0px; bottom: 0px; color: transparent;"
          />
        </div>
      </div>
      <div
        class="c3 c8"
      >
        <div
          class="c5 rounded-lg shadow-lg ratio ratio-7x5"
        >
          <div
            class="d-flex justify-content-center align-items-center"
          >
            <div
              class="c6 spinner-border"
              role="status"
              size="4"
            >
              <span
                class="visually-hidden"
              >
                Loading...
              </span>
            </div>
          </div>
          <img
            alt="Your Design Here"
            class="c7 card-img card-img-fade-in"
            data-nimg="fill"
            decoding="async"
            loading="lazy"
            sizes="100vw"
            src="/_next/image?url=%2Flogo-blank.png&w=3840&q=75"
            srcset="/_next/image?url=%2Flogo-blank.png&w=640&q=75 640w, /_next/image?url=%2Flogo-blank.png&w=750&q=75 750w, /_next/image?url=%2Flogo-blank.png&w=828&q=75 828w, /_next/image?url=%2Flogo-blank.png&w=1080&q=75 1080w, /_next/image?url=%2Flogo-blank.png&w=1200&q=75 1200w, /_next/image?url=%2Flogo-blank.png&w=1920&q=75 1920w, /_next/image?url=%2Flogo-blank.png&w=2048&q=75 2048w, /_next/image?url=%2Flogo-blank.png&w=3840&q=75 3840w"
            style="position: absolute; height: 100%; width: 100%; left: 0px; top: 0px; right: 0px; bottom: 0px; color: transparent;"
          />
        </div>
      </div>
      <div
        class="c3 c9"
      >
        <div
          class="c5 rounded-lg shadow-lg ratio ratio-7x5"
        >
          <div
            class="d-flex justify-content-center align-items-center"
          >
            <div
              class="c6 spinner-border"
              role="status"
              size="4"
            >
              <span
                class="visually-hidden"
              >
                Loading...
              </span>
            </div>
          </div>
          <img
            alt="Your Design Here"
            class="c7 card-img card-img-fade-in"
            data-nimg="fill"
            decoding="async"
            loading="lazy"
            sizes="100vw"
            src="/_next/image?url=%2Flogo-blank.png&w=3840&q=75"
            srcset="/_next/image?url=%2Flogo-blank.png&w=640&q=75 640w, /_next/image?url=%2Flogo-blank.png&w=750&q=75 750w, /_next/image?url=%2Flogo-blank.png&w=828&q=75 828w, /_next/image?url=%2Flogo-blank.png&w=1080&q=75 1080w, /_next/image?url=%2Flogo-blank.png&w=1200&q=75 1200w, /_next/image?url=%2Flogo-blank.png&w=1920&q=75 1920w, /_next/image?url=%2Flogo-blank.png&w=2048&q=75 2048w, /_next/image?url=%2Flogo-blank.png&w=3840&q=75 3840w"
            style="position: absolute; height: 100%; width: 100%; left: 0px; top: 0px; right: 0px; bottom: 0px; color: transparent;"
          />
        </div>
      </div>
      <div
        class="c3 c10"
      >
        <div
          class="c5 rounded-lg shadow-lg ratio ratio-7x5"
        >
          <div
            class="d-flex justify-content-center align-items-center"
          >
            <div
              class="c6 spinner-border"
              role="status"
              size="4"
            >
              <span
                class="visually-hidden"
              >
                Loading...
              </span>
            </div>
          </div>
          <img
            alt="Your Design Here"
            class="c7 card-img card-img-fade-in"
            data-nimg="fill"
            decoding="async"
            loading="lazy"
            sizes="100vw"
            src="/_next/image?url=%2Flogo-blank.png&w=3840&q=75"
            srcset="/_next/image?url=%2Flogo-blank.png&w=640&q=75 640w, /_next/image?url=%2Flogo-blank.png&w=750&q=75 750w, /_next/image?url=%2Flogo-blank.png&w=828&q=75 828w, /_next/image?url=%2Flogo-blank.png&w=1080&q=75 1080w, /_next/image?url=%2Flogo-blank.png&w=1200&q=75 1200w, /_next/image?url=%2Flogo-blank.png&w=1920&q=75 1920w, /_next/image?url=%2Flogo-blank.png&w=2048&q=75 2048w, /_next/image?url=%2Flogo-blank.png&w=3840&q=75 3840w"
            style="position: absolute; height: 100%; width: 100%; left: 0px; top: 0px; right: 0px; bottom: 0px; color: transparent;"
          />
        </div>
      </div>
      <div
        class="c3 c11"
      >
        <div
          class="c5 rounded-lg shadow-lg ratio ratio-7x5"
        >
          <div
            class="d-flex justify-content-center align-items-center"
          >
            <div
              class="c6 spinner-border"
              role="status"
              size="4"
            >
              <span
                class="visually-hidden"
              >
                Loading...
              </span>
            </div>
          </div>
          <img
            alt="Your Design Here"
            class="c7 card-img card-img-fade-in"
            data-nimg="fill"
            decoding="async"
            loading="lazy"
            sizes="100vw"
            src="/_next/image?url=%2Flogo-blank.png&w=3840&q=75"
            srcset="/_next/image?url=%2Flogo-blank.png&w=640&q=75 640w, /_next/image?url=%2Flogo-blank.png&w=750&q=75 750w, /_next/image?url=%2Flogo-blank.png&w=828&q=75 828w, /_next/image?url=%2Flogo-blank.png&w=1080&q=75 1080w, /_next/image?url=%2Flogo-blank.png&w=1200&q=75 1200w, /_next/image?url=%2Flogo-blank.png&w=1920&q=75 1920w, /_next/image?url=%2Flogo-blank.png&w=2048&q=75 2048w, /_next/image?url=%2Flogo-blank.png&w=3840&q=75 3840w"
            style="position: absolute; height: 100%; width: 100%; left: 0px; top: 0px; right: 0px; bottom: 0px; color: transparent;"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
